runner/models.py

```diff
@@ -214,29 +215,40 @@ def clear(self):
         return self
 
     def set_for_restart(self):
-        if self.resume_attempts == 0:
-            return None
-        resume_attempts = self.resume_attempts - 1
         run_id = self.pk
-        execution_id = self.execution_id
-        started = self.started
         output_directory = self.output_directory
         message = self.message
-        if not message:
-            message = {}
-        if "resume" not in message:
-            message["resume"] = []
+        started = self.started
         started_strftime = ""
         exited_strftime = now().strftime("%m/%d/%Y, %H:%M:%S")
         if started:
             started_strftime = started.strftime("%m/%d/%Y, %H:%M:%S")
-        message["resume"].append((started_strftime, exited_strftime, str(execution_id)))
-        self.clear().save()
-        self.message = message
-        self.started = started
-        self.output_directory = output_directory
-        self.resume_attempts = resume_attempts
-        self.save()
+        if not message:
+            message = {}
+        if self.resume_attempts > 0:
+            resume_attempts = self.resume_attempts - 1
+            execution_id = self.execution_id
+            if "resume" not in message:
+                message["resume"] = []
+            message["resume"].append((started_strftime, exited_strftime, str(execution_id)))
+            self.clear().save()
+            self.resume_attempts = resume_attempts
+            self.message = message
+            self.output_directory = output_directory
+            self.save()
+        elif self.restart_attempts > 0:
+            restart_attempts = self.restart_attempts - 1
+            execution_id = None
+            if "restart" not in message:
+                message["restart"] = []
+            message["restart"].append((started_strftime, exited_strftime))
+            self.clear().save()
+            self.restart_attempts = restart_attempts
+            self.message = message
+            self.output_directory = output_directory
+            self.save()
+        else:
+            return None
```

Comment:
This function, I mean. What happens if it returns None?

Reply:
Returning None signifies that the run is out of retries and resumes and should fail. The fail task has the logic here to read None and proceed with failing the run.

https://github.com/mskcc/beagle/blob/master/runner/tasks.py#L515